### PR TITLE
Implemented registry for digester

### PR DESCRIPTION
**Pull request by danw5050**
This will be used to handle custom digester histograms

### CODE REVIEW

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  private val histograms: MutableMap<String, TDigestHistogram<T>> = mutableMapOf()
+
+  /** Creates and returns a new histogram. Histogram is registered within the Registry */
+  override fun newHistogram(
+    name: String,
+    help: String,
+    labelNames: List<String>,
+    quantiles: Map<Double, Double>
+  ): TDigestHistogram<T> {
+    val histogram = TDigestHistogram<T>(name, help, quantiles.keys.toList(), fun() = newDigestFn())
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: The `TDigestHistogram` constructor expects quantiles to be ordered but the order of `quantiles.keys` is non-deterministic.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    return register(histogram)
+  }
+
+  /** Registers a histogram within the Registry */
+  @Synchronized fun register(histogram: TDigestHistogram<T>): TDigestHistogram<T> {
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: This should be private. It's an implementation detail.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    return hash
+  }
+}
+
+/** Default quantiles that can be used to register with a new histogram */
+val defaultQuantiles = mapOf(0.5 to 0.05, 0.75 to 0.02, 0.95 to 0.01, 0.99 to 0.001, 0.999 to 0.0001)
+
+/** TDigestHistogramRegistry registers all TDigestHistograms */
+class TDigestHistogramRegistry <T : TDigest<T>> constructor(
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: There need to be some interface tweaks to match how this will be used in practice.

Current state:
```
    val registry = TDigestHistogramRegistry<VeneurDigest>()

    val histogram = registry.newHistogram(
      name = "request_latency_ms",
      help = "hello",
      labelNames = listOf("foo", "bar"),
      quantiles = mapOf(0.75 to 999.999) // why is this a map?
    )

    // Why does get metric return a nullable?
    histogram.getMetric("x", "y")?.observe(1.23)
```

Expected usage, I think:
```
    val registry = TDigestHistogramRegistry<VeneurDigest>()

    val histogram = registry.newHistogram(
      name = "request_latency_ms",
      help = "hello",
      labelNames = listOf("foo", "bar"),
      quantiles = defaultQuantiles, // or omitted to get defaults automatically
    )

    histogram.metric("x", "y").observe(1.23)
    // or _additional_ possibility
    histogram.metric(mapOf("foo" to "x", "bar" to "y")).observe(1.23)
```

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    labelNames: List<String>,
+    quantiles: Map<Double, Double>
+  ): TDigestHistogram<T> {
+    val histogram = TDigestHistogram<T>(name, help, quantiles.keys.toList(), fun() = newDigestFn())
+    return register(histogram)
+  }
+
+  /** Registers a histogram within the Registry */
+  @Synchronized fun register(histogram: TDigestHistogram<T>): TDigestHistogram<T> {
+    if  (!histograms.containsKey(histogram.getName())) {
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]:
```
  /** Registers a histogram within the Registry */
  @Synchronized
  private fun register(histogram: TDigestHistogram<T>): TDigestHistogram<T> {
    return histograms.getOrPut(histogram.getName()) { histogram }
  }
```

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+/** TDigestHistogramRegistry registers all TDigestHistograms */
+class TDigestHistogramRegistry <T : TDigest<T>> constructor(
+    private val newDigestFn: () -> SlidingWindowDigest<*> =
+        fun() = SlidingWindowDigest(
+            Windower(windowSecs = 30, stagger = 3),
+            fun() = VeneurDigest())
+  ): HistogramRegistry {
+
+  private val histograms: MutableMap<String, TDigestHistogram<T>> = mutableMapOf()
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Java's concurrent maps are richer than what Go offers. `ConcurrentHashMap` is a good fit here too. https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html

`private val histograms = ConcurrentHashMap<String, TDigestHistogram<T>>()`

It also makes the class much simpler:
```
  // ...

  /** Registers a histogram within the Registry */
  private fun register(histogram: TDigestHistogram<T>): TDigestHistogram<T> {
    return histograms.getOrPut(histogram.getName()) { histogram }
  }

  /** Returns a list of all Histograms registered by the Registry */
  fun allHistograms() = histograms.values

  /** Returns a histogram set under the given labels */
  fun getHistogram(labels: String) = histograms[labels]
}
```

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            Windower(windowSecs = 30, stagger = 3), // this does not matter
+            fun() = FakeDigest()))
+
+    val histogram = registry.newHistogram("name", "help", listOf(), mapOf(0.1 to 0.1, 0.2 to 0.2))
+    histogram.record(1.23, "test")
+
+    val firstWindow = histogram.getMetric("test")!!.digest.windows[0].digest as FakeDigest
+
+    // Check that recorded value gets added to the metric
+    Assertions.assertThat(firstWindow.addedValues).isEqualTo(arrayListOf(1.23))
```
Comment (alecholmes) [inline misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: Static import `assertThat` so that you just have `assertThat instead of `Assertions.assertThat`

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    histogram.record(4.56, "test")
+    Assertions.assertThat(firstWindow.addedValues).isEqualTo(arrayListOf(1.23, 4.56))
+
+    // Add a different metric name
+    histogram.record(2.34, "another_test")
+
+    // Check that a new metric gets created for the new key
+    Assertions.assertThat(histogram.metrics.keys.count()).isEqualTo(2)
```
Comment (alecholmes) [inline misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: `assertThat(histogram.metrics.keys).hasSize(2)`

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    val histogram = registry.newHistogram("name", "help", listOf(), mapOf(0.1 to 0.1, 0.2 to 0.2))
+    histogram.record(1.23, "test")
+
+    val firstWindow = histogram.getMetric("test")!!.digest.windows[0].digest as FakeDigest
+
+    // Check that recorded value gets added to the metric
+    Assertions.assertThat(firstWindow.addedValues).isEqualTo(arrayListOf(1.23))
+
+    histogram.record(4.56, "test")
+    Assertions.assertThat(firstWindow.addedValues).isEqualTo(arrayListOf(1.23, 4.56))
```
Comment (alecholmes) [inline misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: Generally don't need specific types. `listOf` is sufficient; don't need `arrayListOf.

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package misk.metrics.digester
```
Comment (alecholmes) [inline misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: Increase the test coverage to at least have parity with the Go implementation.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  /** Returns the number of windows within the histogram */
+  override fun count(vararg labelValues: String): Int {
+    if (metrics.containsKey(key(labelValues))) {
+      return metrics.getValue(key(labelValues)).digest.windows.count()
+    }
+    return  0
+  }
+
+  /** Returns a metric of the histogram. Order of labels matters */
+  fun getMetric(vararg labelValues: String): DigestMetric? {
```
[the remark this answers is not in the document]
Comment (danw5050) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Developers would not be using getMetrics, but how would this make it more straightforward? It makes writing tests easier and it is in the same format as count.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt (hunk 3 above)
Comment (danw5050) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Its a map because prometheus summaries require an error of quantiles estimation (https://prometheus.io/docs/practices/histograms/#errors-of-quantile-estimation). It looks like the prometheus go client provides the algorithm for this:

"A summary would have had no problem calculating the correct percentile value in both cases, at least if it uses an appropriate algorithm on the client side (like the one used by the Go client)."

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    help: String,
+    labelNames: List<String>,
+    quantiles: Map<Double, Double>
+  ): TDigestHistogram<T> {
+    val histogram = TDigestHistogram(name, help, quantiles.keys.toList(), labelNames, fun() = newDigestFn())
+    return register(histogram)
+  }
+
+  /** Registers a histogram within the Registry */
+  @Synchronized private fun register(histogram: TDigestHistogram<T>): TDigestHistogram<T> {
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Not used or needed.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    return histograms.getOrPut(histogram.name) { histogram }
+  }
+
+  /** Returns a list of all Histograms registered by the Registry */
+  fun allHistograms() = histograms.values
+
+  /** Returns a histogram set under the given labels */
+  fun getHistogram(labels: String) = histograms[labels]
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Not used.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  val help: String,
+  val quantiles: List<Double>,
+  private val labelNames: List<String>,
+  private val tDigest: () -> SlidingWindowDigest<T>
+) : Histogram {
+
+  private val metrics = ConcurrentHashMap<Int, DigestMetric>()
+
+  /** Add all labels passed to the histogram */
+  init {
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Remove this. It is unintentionally creating metrics.

Say label names given are `[a, b]`. This creates two metrics with label values `{a->a}` and `{b->b}`

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    return TDigestHistogramRecordMetric(metric)
+  }
+
+  /** Returns the number of windows within the histogram */
+  override fun count(vararg labelValues: String): Int {
+    return metrics.get(key(labelValues))?.digest?.windows?.count() ?: 0
+  }
+
+  /** Returns a metric of the histogram. Order of labels matters */
+  fun getMetric(vararg metric: String): DigestMetric {
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: This doesn't need to be public. Also, Kotlin by convention avoids `get`/`set` prefixes.

```
internal fun metric(vararg metric: String) = metrics[key(metric)]
```

This also changes from `getValue` to normal map get. `getValue` never returns null but your unit tests expect nullable values to come back from this method (`assertThat(histogram.getMetric("label1")).isNotNull()`)

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    return metrics.getValue(key(metric))
+  }
+
+  /** Creates a hash for a list of labels */
+  private fun key(labels: Array<out String>) = Objects.hash(labels.sorted())
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Don't sort here. This is sorting label _values_, which you definitely don't want to do.

The Go implementation is based on a map, which is why it sorts based on the map keys.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+  /** Records a new metric within the histogram */
+  override fun labels(vararg labelValues: String): TDigestHistogramRecordMetric {
+    val metric = metrics.getOrPut(key(labelValues)) {
+      DigestMetric(tDigest(), labelValues.asList())
+    }
+    return TDigestHistogramRecordMetric(metric)
+  }
+
+  /** Returns the number of windows within the histogram */
+  override fun count(vararg labelValues: String): Int {
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Avoid adding public methods just for unit tests. Remove this.

Eventually you'll need something like `fun allMetrics() = metrics.metrics`. Your unit test can use that to assert count.

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+  @Test
+  fun registryDigestWithLabels() {
+    val registry = histogramRegistryFactory()
+
+    val histogram = registry.newHistogram("name", "help", listOf("label1", "label2"),
+        mapOf(0.1 to 0.1, 0.2 to 0.2))
+
+    //Check that passed labels exist
+    assertThat(histogram.getMetric("label1")).isNotNull()
```
Comment (alecholmes) [inline misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: This should not pass. The code should throw an user-friendly exception, as the user must pass a label value for each label name.

`histogram.getMetric()` and `histogram.getMetric("a","b","c","d")` should also throw.

Comment (alecholmes) [inline misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: Code also allows invalid metric names like `foo-bar`.

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt (hunk 9 above)
Comment (alecholmes) [reply misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: Ping. Please take a look at the test cases covered in the Go implementation. There are edge cases that are not tested here, and the implementation needs some fixes for them.

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+    val firstWindow = histogram.getMetric("test").digest.windows[0].digest as FakeDigest
+
+    // Check that recorded value gets added to the metric
+    assertThat(firstWindow.addedValues).isEqualTo(listOf(1.23))
+    metric.observe(4.56)
+
+    // Check that both values are added to the window
+    assertThat(firstWindow.addedValues).isEqualTo(listOf(1.23, 4.56))
+    assertThat(histogram.count("test")).isEqualTo(3)
```
Comment (alecholmes) [inline misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: `assertThat(histogram.allMetrics()).hasSize(3)`

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+    // Add a different metric name
+    histogram.labels("another_test", "another_test_2").observe(2.34)
+
+    // Check that a new metric gets created for the new key
```
Comment (alecholmes) [inline misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: The asserts below don't check that anything is created, only that values are not null.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt (hunk 13 above)
Comment (danw5050) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Do we not want this at all? This is to maintain a similar approach to Prometheus histogram metrics.
```  init {
      labels(*labelNames as Array<out String>)
  } ```

Comment (alecholmes) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: No, because we don't want to mix up label names with label values.

```
// This does not create any metrics _yet_.
// However, it means that every metric derived from it will have exactly two labels
// and the _names_ of those labels will always be label1 and label2.
val hist = registry.newHistogram("yourmetric", "help", listOf("label1", "label2"), mapOf(...))

// This means a metric named "yourmetric" with labels {"label1":"value1", "label2":"value2"} is sent to Prometheus
hist.labels("value1", "value2").observe(1.23)

// This means a metric named "yourmetric" with labels {"label1":"x", "label2":"y"} is sent to Prometheus
hist.labels("x", "y").observe(1.23)
```

Your init here is effectively doing:
```
// Invalid because no observed metric has a _value_ of label1 and also because a value for label2 is omitted.
hist.labels("label1")

// Invalid because no observed metric has a _value_ of label2 and also because a value for label1 is omitted.
hist.labels("label2")
```

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt (hunk 16 above)
Comment (danw5050) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Not sure if it is possible https://stackoverflow.com/questions/13160672/java-getting-an-error-for-implementing-interface-method-with-weaker-access.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt (hunk 15 above)
Comment (danw5050) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: Don't we want labels ["label1",label2"] to have the same key as ["label2",label1"]?

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt (hunk 16 above)
Comment (alecholmes) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: We can change the interface here  since it's misk.metrics.Histogram.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt (hunk 15 above)
Comment (alecholmes) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: No, since the key is based on label _values_ and not label names. So for this you need `Objects.hash(*labels)` (note the important `*` to expand the varargs).

Each `TDigestHistogram` instance represents a metric with a fixed name, a set of labels with fixed names and differing values. They key in the metrics map for this class represents a unique combination of ordered label values.

Let's say there's an instance named `hist` of `TDigestHistogram` with `labelNames == ["a-label", "b-label"]`. Calling
`hist.labels("x", "y")` and `hist.labels("y", "x")` creates two different metric values in Prometheus:
1. `{"a-label":"x", "b-label":"y"}` where the key in your metrics map is `Objects.hash("x", "y")`
2. `{"a-label":"y", "b-label":"x"}` where the key in your metric map here is `Objects.hash("y", "x")`

Even though both examples use `x` and `y` as values for labels, they use them as values for different label names. `Objects.hash("x", "y") != Objects.hash("y, "x")`. If you were to sort label values, they two metric values would end up being incorrectly squished into one metric value.

File: misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt (hunk 18 above)
Comment (danw5050) [reply misk-metrics-digester/src/test/kotlin/misk/metrics/digester/RegistryTest.kt]: hasSize is for lists/arrays etc, count() returns int.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+  val quantiles: List<Double>,
+  private val labelNames: List<String>,
+  private val tDigest: () -> SlidingWindowDigest<T>
+) : Histogram {
+
+  private val metrics = ConcurrentHashMap<Int, DigestMetric>()
+  private val labelRegex = "[a-zA-Z0-9_]*".toRegex()
+
+  /** Records a new metric within the histogram */
+  override fun labels(vararg labelValues: String): TDigestHistogramRecordMetric {
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: This still needs to check that the number label values is the same as the number of label names. This was a comment earlier and needs test coverage.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt (hunk 15 above)
Comment (alecholmes) [reply misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: This still can't be sorted. Let's talk offline to clarify.

File: misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+              fun() = VeneurDigest())
+      )
+    }
+  }
+
+  private val histograms = ConcurrentHashMap<String, TDigestHistogram<T>>()
+  private val metricRegex =  "[a-zA-Z_:][a-zA-Z0-9_:]*".toRegex()
+
+  /** Creates and returns a new histogram. Histogram is registered within the Registry */
+  @Synchronized override fun newHistogram(
```
Comment (alecholmes) [inline misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt]: This shouldn't be synchronized. The use of `histograms` here gives the necessary amount of concurrency control without the overhead of `@Synchronized` being needed.

**Comment by danw5050** [comment]
@alecholmes Yes, there were a couple things I wanted to clarify before continuing. I was waiting until our wed meeting to do that.

**Comment by adrw** [comment]
Rebasing off Master should restore green CI builds.

**Comment by chris-ryan-square** [comment]
If this PR is still relevant, then please fix conflicts, rebase and get a green CI build.
If the PR is no longer relevant please close.

This PR will be closed at the end of January 2021 otherwise.

**Comment by chris-ryan-square** [comment]
Closing request as per previous announcement.
